nibe/connection/modbus.py

```diff
@@ -20,13 +33,39 @@ def __init__(self, heatpump: HeatPump, url, slave_id, conn_options=None):
     async def read_coil(self, coil: Coil, timeout: float = DEFAULT_TIMEOUT) -> Coil:
         logger.debug("Sending read request")
         try:
-            result = await asyncio.wait_for(
-                self._client.read_input_registers(
-                    slave_id=self._slave_id, starting_address=coil.address, quantity=1
-                ),
-                timeout,
-            )
+
+            entity_type, entity_number, entity_count = split_modbus_data(coil)
+
+            with async_timeout.timeout(timeout):
+                if entity_type == 4:
+                    result = await self._client.read_input_registers(
+                        slave_id=self._slave_id,
+                        starting_address=entity_number,
+                        quantity=entity_count,
+                    )
+                elif entity_type == 3:
+                    result = await self._client.read_holding_registers(
```

Comment:
I’d create a mapping from entity type to a method that reads register/input/coil.

Reply:
It get a bit ugly since it's member functions.